PEP 769: Add a 'default' keyword argument to 'attrgetter' and 'itemgetter' (#4179)
python/peps: peps/pep-0769.rst

``default`` keyword argument to the ``attrgetter`` and ``itemgetter``
functions. This addition would allow these functions to return a
specified default value when the targeted attribute or item is missing,
thereby preventing exceptions and simplifying code that handles optional
attributes or items.

Motivation
==========

Currently, ``attrgetter`` and ``itemgetter`` raise exceptions if the
specified attribute or item is absent. This limitation requires
developers to implement additional error handling, leading to more
complex and less readable code.

Introducing a ``default`` parameter would streamline operations involving
optional attributes or items, reducing boilerplate code and enhancing
code clarity.

Rationale
=========

The primary design decision is to introduce a single ``default`` parameter
applicable to all specified attributes or items.

This approach maintains simplicity and avoids the complexity of assigning
individual default values to multiple attributes or items. While some
discussions considered allowing multiple defaults, the increased
complexity and potential for confusion led to favoring a single default

> value for all cases (more about this below in Rejected Ideas).

Reviewer: We could add similar reference here if we want to hyperlink it.

    >>> attrgetter("badname", default="noclass")(C)
    'noclass'
    >>> attrgetter("D", "E", default="noclass")(C)
    (<class '__main__.C.D'>, <class '__main__.C.E'>)
    >>> attrgetter("D", "badname", default="noclass")(C)
    (<class '__main__.C.D'>, 'noclass')
    >>> attrgetter("D.X", default="noclass")(C)
    <class '__main__.C.D.X'>
    >>> attrgetter("D.badname", default="noclass")(C)
    'noclass'

Examples for itemgetter
-----------------------

Current behaviour, no changes introduced::

    >>> obj = ["foo", "bar", "baz"]
    >>> itemgetter(1)(obj)
    'bar'
    >>> itemgetter(5)(obj)
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    IndexError: list index out of range
    >>> itemgetter(1, 0)(obj)
    ('bar', 'foo')
    >>> itemgetter(1, 5)(obj)
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    IndexError: list index out of range

Using ``default``::

    >>> itemgetter(1, default="XYZ")(obj)
    'bar'
    >>> itemgetter(5, default="XYZ")(obj)
    'XYZ'
    >>> itemgetter(1, 0, default="XYZ")(obj)
    ('bar', 'foo')
    >>> itemgetter(1, 5, default="XYZ")(obj)
    ('bar', 'XYZ')

> About Posible Implementations
> -----------------------------

Reviewer: ```suggestion
About Possible Implementations
------------------------------
```

For the case of ``attrgetter`` is quite direct: it implies using
``getattr`` catching a possible ``AttributeError``. So

> ``attrgetter("name", default=XYZ)(obj)`` would be like:

Reviewer: So the following code is shown as code, not blockquote:

```suggestion
``attrgetter("name", default=XYZ)(obj)`` would be like::
```

    try:
        value = getattr(obj, "name")
    except (TypeError, IndexError, KeyError):
        value = XYZ

Note we cannot rely on using ``gettattr`` with a default value, as would
be impossible to distinguish what it returned on each step when an
attribute chain is specified (e.g.
``attrgetter("foo.bar.baz", default=XYZ)``).

For the case of ``itemgetter`` it's not that easy. The more
straightforward way is similar to above, also simple to define and
understand: attempting ``__getitem__`` and catching a possible exception
(any of the three indicated in ``__getitem__`` reference). This way,
``itemgetter(123, default=XYZ)(obj)`` would be equivalent to::

    try:
        value = obj[123]
    except (TypeError, IndexError, KeyError):
        value = XYZ

> However, this would be not as eficient as we'd want for particular cases,

Reviewer: typo:
```suggestion
However, this would be not as efficient as we'd want for particular cases,
```

e.g. using dictionaries where particularly good performance is desired. A
more complex alternative would be::

    if isinstance(obj, dict):
        value = obj.get(123, XYZ)
    else:
        try:
            value = obj[123]
        except (TypeError, IndexError, KeyError):
            value = XYZ

Better performance, more complicated to implement and explain. This is

> the first case in the Open Issues section later.

Reviewer: This could also be a linked reference.

    >>> ig_combined({})
    (1, 2)

However, combining ``itemgetter`` or ``attrgetter`` is a totally new
behaviour very complex to define, not impossible, but beyond the scope of
this PEP.

At the end having multiple default values was deemed overly complex and
potentially confusing, and a single ``default`` parameter was favored for
simplicity and predictability.

Tuple Return Consistency
------------------------

Another rejected proposal was adding a a flag to always return tuple
regardless of how many keys/names/indices were sourced to arguments.
E.g.::

    >>> letters = ["a", "b", "c"]
    >>> itemgetter(1, return_tuple=True)(letters)
    ('b',)
    >>> itemgetter(1, 2, return_tuple=True)(letters)
    ('b', 'c')

This would be of a little help for multiple default values consistency,
but requires further discussion and for sure is out of the scope of this
PEP.

Open Issues
===========

Behaviour Equivalence for ``itemgetter``
----------------------------------------

We need to define how ``itemgetter`` would behave, if just attempt to
access the item and capture exceptions no matter which the object, or
validate first if the object provides a ``get`` method and use it to

> retrieve the item with a default. See examples in the About Posible
> Implementations subsection before.

Reviewer: ```suggestion
retrieve the item with a default. See examples in the About Possible
Implementations subsection above.
```
Or we could add a reference?

For example, add this above:

```diff
+.. _PEP 769 About Possible Implementations:
 
 About Posible Implementations
 -----------------------------
```

And then:

```suggestion
retrieve the item with a default. See examples in the `About Possible
Implementations <PEP 769 About Possible Implementations_>`__ subsection above.
```

confusing if some object that is not a dictionary but provides a ``get``
method is used. Alternatively, we could call ``.get`` *only* if the
object is an instance of ``dict``.

In any case, a desirable situation is that we do *not* affect performance
at all if the ``default`` is not triggered. Checking for ``.get`` would
get the default faster in case of dicts, but implies doing a verification
in all cases. Using the try/except model would make it not as fast as it
could in the case of dictionaries, but would not introduce delays if the
default is not triggered.

Add a Default to ``getitem``
----------------------------

It was proposed that we could also enhance ``getitem``, as part of the of
this PEP, adding ``default`` also to it.

This will not only improve ``getitem`` itself, but we would also gain
internal consistency in the ``operator`` module and in comparison with
the ``getattr`` builtin function that also has a default.

The definition could be as simple as the try/except proposed above, so
doing ``getitem(obj, name, default)`` would be equivalent to::

    try:
        result = obj[name]
    except (TypeError, IndexError, KeyError):

>         resutl = default

Reviewer: ```suggestion
        result = default
```

> (however see previous open issue about special case for dictionaries)

Reviewer: ```suggestion
(However see previous open issue about special case for dictionaries.)
```

functions. This addition would allow these functions to return a
specified default value when the targeted attribute or item is missing,
thereby preventing exceptions and simplifying code that handles optional
attributes or items.

Motivation
==========

Currently, ``attrgetter`` and ``itemgetter`` raise exceptions if the
specified attribute or item is absent. This limitation requires
developers to implement additional error handling, leading to more
complex and less readable code.

Introducing a ``default`` parameter would streamline operations involving
optional attributes or items, reducing boilerplate code and enhancing
code clarity.

Rationale
=========

The primary design decision is to introduce a single ``default`` parameter
applicable to all specified attributes or items.

This approach maintains simplicity and avoids the complexity of assigning
individual default values to multiple attributes or items. While some
discussions considered allowing multiple defaults, the increased
complexity and potential for confusion led to favoring a single default
value for all cases (more about this below in `Rejected Ideas

> <PEP 769 Rejected Ideas>`__).

Reviewer: We need an extra underscore here:
```suggestion
<PEP 769 Rejected Ideas_>`__).
```

    except (TypeError, IndexError, KeyError):
        value = XYZ

Note we cannot rely on using ``gettattr`` with a default value, as would
be impossible to distinguish what it returned on each step when an
attribute chain is specified (e.g.
``attrgetter("foo.bar.baz", default=XYZ)``).

For the case of ``itemgetter`` it's not that easy. The more
straightforward way is similar to above, also simple to define and
understand: attempting ``__getitem__`` and catching a possible exception
(any of the three indicated in ``__getitem__`` reference). This way,
``itemgetter(123, default=XYZ)(obj)`` would be equivalent to::

    try:
        value = obj[123]
    except (TypeError, IndexError, KeyError):
        value = XYZ

However, this would be not as efficient as we'd want for particular cases,
e.g. using dictionaries where particularly good performance is desired. A
more complex alternative would be::

    if isinstance(obj, dict):
        value = obj.get(123, XYZ)
    else:
        try:
            value = obj[123]
        except (TypeError, IndexError, KeyError):
            value = XYZ

Better performance, more complicated to implement and explain. This is

> the first case in the `Open Issues <PEP 769 Open Issues>`__ section later.

Reviewer: ```suggestion
the first case in the `Open Issues <PEP 769 Open Issues_>`__ section later.
```

behaviour very complex to define, not impossible, but beyond the scope of
this PEP.

At the end having multiple default values was deemed overly complex and
potentially confusing, and a single ``default`` parameter was favored for
simplicity and predictability.

Tuple Return Consistency
------------------------

Another rejected proposal was adding a a flag to always return tuple
regardless of how many keys/names/indices were sourced to arguments.
E.g.::

    >>> letters = ["a", "b", "c"]
    >>> itemgetter(1, return_tuple=True)(letters)
    ('b',)
    >>> itemgetter(1, 2, return_tuple=True)(letters)
    ('b', 'c')

This would be of a little help for multiple default values consistency,
but requires further discussion and for sure is out of the scope of this
PEP.

.. _PEP 769 Open Issues:

Open Issues
===========

Behaviour Equivalence for ``itemgetter``
----------------------------------------

We need to define how ``itemgetter`` would behave, if just attempt to
access the item and capture exceptions no matter which the object, or
validate first if the object provides a ``get`` method and use it to
retrieve the item with a default. See examples in the `About Possible

> Implementations <PEP 769 About Possible Implementations>`__ subsection

Reviewer: ```suggestion
Implementations <PEP 769 About Possible Implementations_>`__ subsection
```
